mod: Beginner Guide: rewrote the gamma correction section (#349)
IrisShaders/DocsPage: src/content/docs/current/Guides/Your First Shader/1_composite.mdx

---

> title: Your First Effect

Reviewer: ```suggestion
title: Your First Post Processing Effect
```

Author: I didn't do this because of

<img width="410" alt="image" src="https://github.com/user-attachments/assets/6408a670-b265-43fd-a84c-8f1ab874a7b2" />

That's on a 14-in laptop in full screen, which I think is how a decent amount of people will view the tutorial. "Your First Shader Effect" could be a good alternative, or you could just accept this

description: Write a basic post processing effect in the `composite` pass.
sidebar:

>   label: Your First Effect

Reviewer: ```suggestion
  label: Your First Post Processing Effect
```

#version 330 compatibility
```

> Every shader's first line must be a version declaration. GLSL has two profiles: `compatibility` and `core`. Iris is better at patching the `compatibility` profile, so in the context of Minecraft there's no benefit to using `core`.

Reviewer: ```suggestion
Every shader's first line must be a version declaration. GLSL has two profiles: `compatibility` and `core`. Iris is better at patching the `compatibility` profile, so in this case there is no benefit to using `core`.
```

```glsl
out vec2 texcoord;
```

> This is a variable declaration, but a special one. The `out` keyword means that the value will be passed to the fragment shader. The fragment shader can then have a corresponding `in` declaration which allows it to recieve this value. In older versions of GLSL, both `in` and `out` were replaced with the `varying` keyword, which people sometimes still call it.

Reviewer: ```suggestion
This is a variable declaration, but a special one. The `out` keyword means that the value will be passed to the fragment shader. The fragment shader can then have a corresponding `in` declaration which allows it to receive this value. In older versions of GLSL, both `in` and `out` were represented with the `varying` keyword, and variables passed between programs are often referred to as 'varyings'.
```

void main() {
```

> The `main` function is the code that is run when the shader is invoked, just like in C.

Reviewer: ```suggestion
The `main` function is invoked when the shader program runs, just like in C.
```

```glsl
gl_Position = ftransform();
```

> This transforms the position of the vertex from model space to clip space. The `ftransform` function is actually deprecated but Iris patches it to the relevant modern code. For more information, see the Iris docs on [coordinate spaces](/current/how-to/coordinate_spaces).

Reviewer: ```suggestion
This transforms the position of the vertex from model space to clip space. The `ftransform` function is actually deprecated, but Iris patches it to the relevant modern code. For more information, see the docs on [coordinate spaces](/current/how-to/coordinate_spaces).
```

void main() {
```

The `main` function is invoked when the shader program runs, just like in C.

```glsl
gl_Position = ftransform();
```

This transforms the position of the vertex from model space to clip space. The `ftransform` function is actually deprecated, but Iris patches it to the relevant modern code. For more information, see the docs on [coordinate spaces](/current/how-to/coordinate_spaces).

```glsl
texcoord = (gl_TextureMatrix[0] * gl_MultiTexCoord0).xy;
```

This gives us the 'texture coordinate' of the current vertex. This is more commonly known as the 'UV', and it is used so that the fragment shader knows where on the screen it is. These texture coordinates range from (0, 0) at the bottom left of the texture to (1, 1) at the top right.

:::tip[Swizzling]
In the previous line of code, you might have noticed some weird syntax: `.xy`. This is an operation unique to shading languages known as *swizzling*. You can read more about swizzling [here](https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)#Swizzling).
:::

> Let's open `composite.fsh`. This is the *fragment shader*, meaning it runs for every pixel on the screen.

Reviewer: A fragment shader doesn't necessarily run for every pixel on the screen (i.e gbuffers)
```suggestion
Let's open `composite.fsh`. This is the *fragment shader*, and since `composite` is a *full screen pass*, it runs for every pixel on the screen.
```

color = texture(colortex0, texcoord);
```

This reads the value in `colortex0` at position `texcoord` and stores it in `color`. For more info, see [the OpenGL docs](https://registry.khronos.org/OpenGL-Refpages/gl4/html/texture.xhtml).

## Making It Grayscale

> A color is grayscale when the r, g, and b components all have the same value. We can compute this value by taking the dot product of the original color and a `vec3(1.0/3.0)`. This is mathematically equivalent to multiplying each channel by 1/3 and then summing the products together. If we set the r, g, and b channels of `color` to this value, the resulting image will be converted to grayscale.

Reviewer: ```suggestion
A color is grayscale when the r, g, and b components all have the same value. We can compute this value by taking the dot product of the original color and a `vec3(1.0/3.0)`. This is mathematically equivalent to multiplying each channel by 1/3 and then summing the products together (averaging them). If we set the r, g, and b channels of `color` to this value, the resulting image will be converted to grayscale.
```

This reads the value in `colortex0` at position `texcoord` and stores it in `color`. For more info see [the OpenGL docs](https://registry.khronos.org/OpenGL-Refpages/gl4/html/texture.xhtml).

## Making It Grayscale

> A color is grayscale when the r, g, and b components all have the same value. We can compute this value by taking the dot product of the original color and a `vec3(1.0/3.0)`. This is mathematically equivalent to multiplying each channel by 1/3 and then summing the products together. If we set all three channels for color to this value, our game will be converted to grayscale.

Reviewer: ```suggestion
A color is grayscale when the r, g, and b components all have the same value. We can compute this value by taking the dot product of the original color and a `vec3(1.0/3.0)`. This is mathematically equivalent to multiplying each channel by 1/3 and then summing the products together. If we set the r, g, and b channels of `color` to this value, the resulting image will be converted to grayscale.
```
